extends TabWindow

@onready var list = $Control/scenes/list

func _ready() -> void:
	window_hide()
	await get_tree().process_frame
	glob.reset_menu_type(list, "list2")
	$CanvasLayer.hide()
	glob.un_occupy(list, &"menu")
	glob.un_occupy(list, "menu_inside")

	border_rect_1 = ColorRect.new()
	border_rect_1.color.a = 0
	border_rect_1.mouse_default_cursor_shape = Control.CURSOR_HSIZE
	add_child(border_rect_1)

	border_rect_2 = ColorRect.new()
	border_rect_2.color.a = 0
	border_rect_2.mouse_default_cursor_shape = Control.CURSOR_HSIZE
	add_child(border_rect_2)

	border_console = ColorRect.new()
	border_console.color.a = 0
	border_console.mouse_default_cursor_shape = Control.CURSOR_VSIZE
	add_child(border_console)
	

	repos()


func repos():
	border_rect_1.position = Vector2($Control/scenes.size.x - border_hit, $Control/scenes.global_position.y)
	border_rect_1.size = Vector2(border_hit * 2, $Control/scenes.size.y)
	border_rect_2.position = Vector2($Control/CodeEdit.size.x + $Control/CodeEdit.position.x, $Control/scenes.global_position.y)
	border_rect_2.size = Vector2(border_hit, $Control/scenes.size.y)
	var code_pos = $Control/CodeEdit.global_position
	var console_node = $Control/console
	var top = console_node.global_position.y
	border_console.position = Vector2(console_node.global_position.x, top - border_hit)
	border_console.size = Vector2(console_node.size.x - 10, border_hit * 2)
	$Control/console.position.x = code_pos.x
	$Control/console.size.x = $Control/CodeEdit.size.x
	$Control/CodeEdit.size.y = $Control/scenes.size.y - $Control/console.size.y


var border_rect_1: ColorRect
var border_rect_2: ColorRect


func _window_hide():
	process_mode = Node.PROCESS_MODE_DISABLED
	hide()
	glob.reset_menu_type(list, "list2")
	$CanvasLayer.hide()
	glob.un_occupy(list, &"menu")
	glob.un_occupy(list, "menu_inside")


@onready var base_size = $Control.size
func _window_show():
	process_mode = Node.PROCESS_MODE_ALWAYS
	$Control.position.y = glob.space_begin.y
	$Control.size.y = glob.window_size.y - $Control.position.y
	$CanvasLayer.show()
	show()
	$Control/scenes/list.menu_show($Control/scenes/list.global_position)
	prev_win = Vector2()
	tick()
	reload_scenes()
	await get_tree().process_frame
	$Control/scenes/list.update_children_reveal()


func _process(delta: float) -> void:
	tick()
	var bot = 0
	var top = 0
	var vbar: VScrollBar = $Control/console.get_v_scroll_bar()
	if vbar.value < vbar.max_value - vbar.page - 5:
		bot = $Control/console.get_global_rect().end.y - 5
	if vbar.value > 5:
		top = $Control/console.global_position.y+30
	$Control/console.set_instance_shader_parameter("extents", Vector4(top, bot, 0, 0))

# ---- division settings ----
var division_ratio: Array[float] = [0.2, 0.6]
var min_scenes_size: float = 160.0
var max_scenes_size: float = 300.0
var min_game_size: float = 60.0
var max_game_size: float = 300.0
var prev_win: Vector2 = Vector2()
var border_hit = 10.0
var _dragging: int = -1
var _drag_anchor: float = 0.0
var code_hidden: bool = false
var max_game_size_ncode: float = 1800.0


func set_code_hidden(hidden: bool) -> void:
	code_hidden = hidden
	$Control/CodeEdit.visible = not hidden
	tick(true)
	ui.move_mouse(get_global_mouse_position())


func tick(force: bool = false) -> void:
	if not visible:
		return
	handle_division_drag()

	var win: float = glob.window_size.x
	if 1:#prev_win != glob.window_size or _dragging != -1 or force:
		$Control.position.y = glob.space_begin.y
		$Control.size.y = glob.window_size.y - $Control.position.y

		var scenes_w = clamp(win * division_ratio[0], min_scenes_size, max_scenes_size)
		var codeedit_target = min(win * division_ratio[1], glob.window_size.x - min_game_size - $Control/scenes.size.x)
		var codeedit_w = 0.0 if code_hidden else codeedit_target

		var game_w = win - scenes_w - codeedit_w
		game_w = clamp(game_w, min_game_size, max_game_size_ncode if code_hidden else max_game_size)

		if not code_hidden:
			if codeedit_w < 0.0:
				var deficit = -codeedit_w
				game_w = max(min_game_size, game_w - deficit)
				codeedit_w = max(0.0, win - scenes_w - game_w)
				if codeedit_w < 0.0:
					codeedit_w = 0.0
					game_w = max(0.0, win - scenes_w)

		$Control/scenes.size.x = scenes_w
		$Control/CodeEdit.position.x = scenes_w
		$Control/CodeEdit.size.x = codeedit_w

		$Control/view.position.x = scenes_w + codeedit_w
		$Control/view.size.x = game_w
		#print($Control/view.get_global_rect().end.x > glob.window_size.x)

		list.set_menu_size(
			($Control/scenes.size.x - list.position.x * 2 + 3) / list.scale.x,
			($Control/scenes.size.y - list.position.y - 10) / list.scale.y
		)

		if code_hidden:
			$Control/scenes.size.x += 2
		prev_win = glob.window_size
		repos()


func reload_scenes():
	$Control/scenes/list.show_up({"hi": "haha"})
	#if not current_lua_env:
	#	if not received_texts:
	#		$Control/CodeEdit.text = "-- Create your scene in the Scenes menu."
	#	else:
	#		$Control/CodeEdit.text = "-- Select your scene from the Scenes menu."
	#await get_tree().process_frame
	#if !is_instance_valid(last_button):
		#if last_hint in $Control/scenes/list.button_by_hint:
			#_on_list_child_button_release($Control/scenes/list.button_by_hint[last_hint])
		#elif received_texts and $Control/scenes/list._contained:
			#_on_list_child_button_release($Control/scenes/list._contained[0])

var border_console: ColorRect
var _dragging_console: bool = false
var _drag_anchor_y_console: float = 0.0
var console_min_height: float = 50.0
var console_max_height: float = 500.0




func handle_top_drag() -> void:
	if not visible:
		return

	var ctrl = $Control
	var codeedit = ctrl.get_node("CodeEdit")
	var console = ctrl.get_node("console")
	if console == null or codeedit == null:
		return

	var mp = get_global_mouse_position()
	var border_y = console.global_position.y           # border between them

	# Start drag
	if glob.mouse_just_pressed and abs(mp.y - border_y) <= border_hit and mp.x > console.global_position.x and mp.x < console.get_global_rect().end.x - 10:
		_dragging_console = true
		_drag_anchor_y_console = mp.y - border_y
	elif not glob.mouse_pressed and _dragging_console:
		_dragging_console = false

	# While dragging
	if _dragging_console and glob.mouse_pressed:
		var new_border_y = mp.y - _drag_anchor_y_console
		var local_y = new_border_y - ctrl.global_position.y

		# Total vertical span for both areas
		var total_height = ctrl.size.y
		var new_codeedit_height = clamp(local_y, 100.0, total_height - console_min_height)
		var new_console_height = clamp(total_height - new_codeedit_height, console_min_height, console_max_height)

		# Apply sizes and positions in parent's space
		codeedit.position.y = 0
		codeedit.size.y = new_codeedit_height

		console.position.y = new_codeedit_height
		console.size.y = new_console_height

		# Update visual drag bar
		border_console.position = Vector2(console.global_position.x, console.global_position.y - border_hit)
		border_console.size = Vector2(console.size.x, border_hit * 2)
		repos()




func handle_division_drag() -> void:
	handle_top_drag()
	if not visible:
		return

	var win = glob.window_size.x
	if win <= 0.0:
		return

	var scenes_w = $Control/scenes.size.x
	var code_target = win * division_ratio[1]
	var game_w = clamp(win - scenes_w - code_target, min_game_size, max_game_size)
	var code_w = max(0.0, win - scenes_w - game_w)

	var border1 = scenes_w
	var border2 = scenes_w + code_w

	var mp = get_global_mouse_position()
	var ctrl = $Control
	var in_y = mp.y >= ctrl.position.y and mp.y <= (ctrl.position.y + ctrl.size.y)

	if glob.mouse_just_pressed and in_y:
		if abs(mp.x - border1) <= border_hit:
			_dragging = 0
			_drag_anchor = border1 - mp.x
		elif mp.x - border2 <= border_hit and mp.x > border2:
			_dragging = 1
			_drag_anchor = border2 - mp.x
	#var new_xs = clamp(mp.x + _drag_anchor, 0.0, win)
	#var new_sceness = clamp(new_xs, min_scenes_size, max_scenes_size)
	#var desired_codes = max(0.0, border2 - new_sceness)
#
	var new_games = clamp(win - $Control/scenes.size.x - $Control/CodeEdit.size.x, min_game_size, max_game_size)
	var new_codes = max(0.0, win - $Control/scenes.size.x - new_games)
	division_ratio[1] = new_codes / win

	if glob.mouse_pressed and _dragging != -1:
		var new_x = clamp(mp.x + _drag_anchor, 0.0, win)

		if _dragging == 0:
			var new_scenes = clamp(new_x, min_scenes_size, max_scenes_size)
			var desired_code = max(0.0, border2 - new_scenes)

			var new_game = clamp(win - new_scenes - desired_code, min_game_size, max_game_size)
			var new_code = max(0.0, win - new_scenes - new_game)

			division_ratio[0] = new_scenes / win
			division_ratio[1] = new_code / win

		elif _dragging == 1:
			var desired_code = max(0.0, new_x - scenes_w)

			var new_game = clamp(win - scenes_w - desired_code, min_game_size, max_game_size)
			var new_code = max(0.0, win - scenes_w - new_game)

			division_ratio[0] = scenes_w / win
			division_ratio[1] = new_code / win

	if not glob.mouse_pressed and _dragging != -1:
		_dragging = -1
		_drag_anchor = 0.0
	
	#division_ratio[0] = $Control/scenes.size.x / win
	#division_ratio[1] = $Control/CodeEdit.size.x / win

	$Control/view/Label.resize()


func _on__released() -> void:
	set_code_hidden(!code_hidden)


# simplified “+” button
@onready var plus = $Control/scenes/plus
func _on_plus_released() -> void:
	plus.block_input()
	var a = await ui.splash_and_get_result("dataset_create", plus, null, false)
	await get_tree().process_frame
	var m = func(): return glob.mouse_pressed
	while m.call():
		await get_tree().process_frame
	plus.unblock_input()
